src/phy/sys/bpf.rs

```diff
@@ -0,0 +1,122 @@
+use std::io;
+use std::os::unix::io::{AsRawFd, RawFd};
+
+use libc;
+
+use super::{ifreq, ifreq_for};
+
+/// set interface
+const BIOCSETIF: libc::c_ulong = 0x8020426c;
+/// get buffer length
+const BIOCGBLEN: libc::c_ulong = 0x40044266;
+/// set immediate/nonblocking read
+const BIOCIMMEDIATE: libc::c_ulong = 0x80044270;
+
+// TODO: check if this is same for OSes other than macos
+const BPF_HDRLEN: usize = 18;
+
+macro_rules! try_ioctl {
+    ($fd:expr,$cmd:expr,$req:expr) => {
+        unsafe {
+            if libc::ioctl($fd, $cmd, $req) == -1 {
+                return Err(io::Error::last_os_error());
+            }
+        }
+    };
+}
+
+#[derive(Debug)]
+pub struct BpfDevice {
+    fd: libc::c_int,
+    ifreq: ifreq,
+}
+
+impl AsRawFd for BpfDevice {
+    fn as_raw_fd(&self) -> RawFd {
+        self.fd
+    }
+}
+
+impl BpfDevice {
+    pub fn new(name: &str) -> io::Result<BpfDevice> {
+        let fd = unsafe {
+            let mut fd = -1;
```

Comment:
Would be more "rusty" to use an option here instead of using `-1` as the "unpopulated" value.